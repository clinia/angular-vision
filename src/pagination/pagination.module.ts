import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgCviPagination } from './pagination';

export { NgCviPagination } from './pagination';

@NgModule({
  declarations: [NgCviPagination],
  entryComponents: [NgCviPagination],
  exports: [NgCviPagination],
  imports: [CommonModule],
})
export class NgCviPaginationModule {}
