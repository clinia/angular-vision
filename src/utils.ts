export const bem = (widgetName: string) => {
  return (element?: string, subElement?: string): string => {
    let cssClass = `cvi-${widgetName}`;
    if (element) {
      cssClass += `-${element}`;
    }
    if (subElement) {
      cssClass += `--${subElement}`;
    }
    return cssClass;
  };
};

export const parseNumberInput = (input?: number | string): number => {
  return typeof input === 'string' ? parseInt(input, 10) : input;
};

export const noop = (...args: any[]): void => {};

export const capitalize = (s: string): string => {
  return s.charAt(0).toUpperCase() + s.slice(1);
};
