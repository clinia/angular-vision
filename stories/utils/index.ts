export { default as meta } from './meta';
export { default as wrapWithRecords } from './wrap-with-records';
