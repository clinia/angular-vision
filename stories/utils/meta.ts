import { moduleMetadata } from '@storybook/angular';

import { NgCviModule } from '@clinia/angular-vision';

const meta = moduleMetadata({
  imports: [NgCviModule.forRoot()],
});

export default meta;
