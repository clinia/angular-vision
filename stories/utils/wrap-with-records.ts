import { Component } from '@angular/core';
import clinia from 'clinia/lite';

interface Helper {
  search: Function;
}

interface WrapWithRecordsParams {
  template: string;
  styles?: string[];
  searchParameters?: {};
  methods?: {};
  searchFunction?: (helper: Helper) => void;
  searchClient?: {};
  // TODO: update with Vision.js types
  insightsClient?: (method: string, payload: object) => void;
  indexName?: string;
  appId?: string;
  apiKey?: string;
  records?: string;
  routing?: boolean | {};
}

const defaultRecords = `
<ng-template let-records="records">
  <ol class="playground-records">
    <li
      *ngFor="let record of records"
      class="record playground-records-item"
      id="record-{{record.id}}"
    >
      {{record.name}}
    </li>
  </ol>
</ng-template>
`;

export default ({
  template,
  styles = [],
  searchParameters = {},
  methods = {},
  searchFunction,
  insightsClient,
  indexName = 'health_facility',
  appId = 'demo-pharmacies',
  apiKey = 'KcLxBhVFP8ooPgQODlAxWqfNg657fTz9',
  records = defaultRecords,
  routing,
}: WrapWithRecordsParams) => {
  @Component({
    selector: 'cvi-app',
    styles,
    template: `
      <cvi-vision [config]="config">
        <div class="cvi-container cvi-container-preview">
          ${template}
        </div>
        <div class="cvi-container cvi-container-playground">
          <div class="panel-left">
          </div>
          <div class="panel-right">
            <cvi-search-box placeholder="Search into pharmacies"></cvi-search-box>
            <cvi-records>
              ${records}
            </cvi-records>
            <cvi-pagination [totalPages]="20"></cvi-pagination>
          </div>
        </div>
      </cvi-vision>
    `,
  })
  class AppComponent {
    config = {
      searchClient: clinia(appId, apiKey),
      insightsClient,
      indexName,
      searchFunction,
      searchParameters: {
        perPage: 3,
        ...searchParameters,
      },
      routing,
    };

    constructor() {
      Object.keys(methods).forEach(methodName => {
        this[methodName] = methods[methodName];
      });
    }
  }

  return AppComponent;
};
