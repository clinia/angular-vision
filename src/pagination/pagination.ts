const range = require('lodash/range');
import { Component, Input, Inject, forwardRef } from '@angular/core';
import { connectPagination } from '@clinia/vision/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgCviVision } from '../vision/vision';
import { parseNumberInput, noop } from '../utils';

@Component({
  selector: 'cvi-pagination',
  template: `
    <div [ngClass]="[cx(), state.numPages <= 1 ? cx('', 'noRefinement') : '']">
      <ul [class]="cx('list')">
        <li
          *ngIf="showFirst"
          (click)="refine($event, 0)"
          [class]="
            cx('item') +
            ' ' +
            cx('item', 'firstPage') +
            (state.currentRefinement === 0 ? ' ' + cx('item', 'disabled') : '')
          "
        >
          <a
            [href]="state.createURL(0)"
            [class]="cx('link')"
          >
            ‹‹
          </a>
        </li>

        <li
          *ngIf="showPrevious"
          (click)="refine($event, state.currentRefinement - 1)"
          [class]="
            cx('item') +
            ' ' +
            cx('item', 'previousPage') +
            (state.currentRefinement === 0 ? ' ' + cx('item', 'disabled') : '')
          "
        >
          <a
            [href]="state.createURL(state.currentRefinement - 1)"
            [class]="cx('link')"
          >
            ‹
          </a>
        </li>

        <li
          [class]="
            cx('item') +
            ' ' +
            cx('item', 'page') +
            (state.currentRefinement === page ? ' ' + cx('item', 'selected') : '')
          "
          *ngFor="let page of pages"
          (click)="refine($event, page)"
        >
          <a
            [class]="cx('link')"
            [href]="state.createURL(page)"
          >
            {{page + 1}}
          </a>
        </li>

        <li
          *ngIf="showNext"
          (click)="refine($event, state.currentRefinement + 1)"
          [class]="
            cx('item') +
            ' ' +
            cx('item', 'nextPage') +
            (state.currentRefinement + 1 === state.numPages ? ' ' + cx('item', 'disabled') : '')
          "
        >
          <a
            [href]="state.createURL(state.currentRefinement + 1)"
            [class]="cx('link')"
          >
            ›
          </a>
        </li>

        <li
          *ngIf="showLast"
          (click)="refine($event, state.numPages - 1)"
          [class]="
            cx('item') +
            ' ' +
            cx('item', 'lastPage') +
            (state.currentRefinement + 1 === state.numPages ? ' ' + cx('item', 'disabled') : '')
          "
        >
          <a
            [href]="state.createURL(state.numPages - 1)"
            [class]="cx('link')"
          >
            ››
          </a>
        </li>
      </ul>
    </div>
  `,
})
export class NgCviPagination extends BaseWidget {
  // rendering options
  @Input() public showFirst = true;
  @Input() public showLast = true;
  @Input() public showPrevious = true;
  @Input() public showNext = true;
  @Input() public padding: number | string = 3;

  // instance options
  @Input() public totalPages?: number | string;

  public state = {
    createURL: noop,
    currentRefinement: 0,
    total: 0,
    numPages: 0,
    refine: noop,
  };

  get pages() {
    const { numPages, currentRefinement } = this.state;

    const pagesArray = Array.apply(null, { length: numPages }).map(
      Number.call,
      Number
    );

    const pagesPadding =
      typeof this.padding === 'string'
        ? parseInt(this.padding, 10)
        : this.padding;

    if (pagesPadding && pagesPadding > 0) {
      // should not display pages that does not exists
      if (numPages < pagesPadding * 2 + 1) {
        return pagesArray;
      }

      const minDelta = currentRefinement - pagesPadding - 1;
      const maxDelta = currentRefinement + pagesPadding + 1;

      if (minDelta < 0) {
        return range(0, currentRefinement + pagesPadding + Math.abs(minDelta));
      }

      if (maxDelta > numPages) {
        return range(
          currentRefinement - pagesPadding - (maxDelta - numPages),
          numPages
        );
      }

      return range(
        currentRefinement - pagesPadding,
        currentRefinement + pagesPadding + 1
      );
    }

    return pagesArray;
  }

  constructor(
    @Inject(forwardRef(() => NgCviVision))
    public visionParent: NgCviVision
  ) {
    super('Pagination');
  }

  public ngOnInit() {
    this.createWidget(connectPagination, {
      maxPages: parseNumberInput(this.totalPages),
    });
    super.ngOnInit();
  }

  public refine(event: MouseEvent, page: number) {
    event.stopPropagation();
    event.preventDefault();

    if (
      page < 0 ||
      page === this.state.currentRefinement ||
      page >= this.state.numPages
    ) {
      return;
    }

    this.state.refine(page);
  }
}
