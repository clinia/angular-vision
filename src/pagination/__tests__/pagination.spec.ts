import { createRenderer } from '../../../test/test-renderer';
import { NgCviPagination } from '../pagination';

import { bem } from '../../utils';

const cx = bem('Pagination');

const defaultState = {
  createURL: jest.fn(),
  currentRefinement: 0,
  total: 100,
  numPages: 20,
  refine: jest.fn(),
};

const render = createRenderer({
  defaultState,
  template: '<cvi-pagination></cvi-pagination>',
  TestedWidget: NgCviPagination,
});

describe('Pagination', () => {
  it('renders markup without state', () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it('renders with pages in state', () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it('renders does not uses pagesPadding when numPages < pagesPadding * 2 + 1', () => {
    const fixture = render({ numPages: 5 });
    const pages = fixture.debugElement.nativeElement.querySelectorAll(
      `.${cx('item', 'page')}`
    );

    expect(pages.length).toBe(5);
    expect(fixture).toMatchSnapshot();
  });

  it('should refine when clicking a page', () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    const el = fixture.debugElement.nativeElement;
    const [, secondPage] = el.querySelectorAll(`.${cx('item', 'page')}`);
    secondPage.click();

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith(1);

    fixture.componentInstance.testedWidget.state.currentRefinement = 3;
    fixture.detectChanges();

    const previous = el.querySelector(`.${cx('item', 'previousPage')}`);
    previous.click();

    expect(refine).toHaveBeenLastCalledWith(2);

    const next = el.querySelector(`.${cx('item', 'nextPage')}`);
    next.click();

    expect(refine).toHaveBeenLastCalledWith(4);
  });

  it('should display last page button', () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    fixture.componentInstance.testedWidget.showLast = true;
    fixture.detectChanges();

    const lastPage = fixture.debugElement.nativeElement.querySelector(
      `.${cx('item', 'lastPage')}`
    );

    lastPage.click();

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith(defaultState.numPages - 1);
  });

  it("should not refine when there's no other pages", () => {
    const refine = jest.fn();
    const fixture = render({ refine, numPages: 1 });

    fixture.componentInstance.testedWidget.showLast = true;
    fixture.detectChanges();

    const firstPage = fixture.debugElement.nativeElement.querySelector(
      `.${cx('item', 'firstPage')}`
    );
    const previousPage = fixture.debugElement.nativeElement.querySelector(
      `.${cx('item', 'previousPage')}`
    );
    const nextPage = fixture.debugElement.nativeElement.querySelector(
      `.${cx('item', 'nextPage')}`
    );
    const lastPage = fixture.debugElement.nativeElement.querySelector(
      `.${cx('item', 'lastPage')}`
    );

    firstPage.click();
    previousPage.click();
    nextPage.click();
    lastPage.click();

    expect(refine).not.toHaveBeenCalled();
  });
  it('should not add cvi-Pagination--noRefinement CSS class on root when numPages > 1', () => {
    const fixture = render({ numPages: 2 });
    expect(
      fixture.debugElement.nativeElement.querySelector(
        'cvi-pagination > .cvi-Pagination--noRefinement'
      )
    ).toBeFalsy();
    expect(fixture).toMatchSnapshot();
  });
  it('should add cvi-Pagination--noRefinement CSS class on root when numPages === 1', () => {
    const fixture = render({ numPages: 1 });
    expect(
      fixture.debugElement.nativeElement.querySelector(
        'cvi-pagination > .cvi-Pagination--noRefinement'
      )
    ).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });
  it('should add cvi-Pagination--noRefinement CSS class on root when numPages === 0', () => {
    const fixture = render({ numPages: 0 });
    expect(
      fixture.debugElement.nativeElement.querySelector(
        'cvi-pagination > .cvi-Pagination--noRefinement'
      )
    ).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });
});
