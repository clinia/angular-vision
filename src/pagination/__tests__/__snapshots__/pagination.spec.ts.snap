// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders markup without state 1`] = `
<ng-component
  testedWidget={[Function NgCviPagination]}
>
  <cvi-vision>
    <cvi-pagination>
      <div
        class="cvi-Pagination cvi-Pagination--noRefinement"
        ng-reflect-ng-class="cvi-Pagination,cvi-Pagination-"
      >
        <ul
          class="cvi-Pagination-list"
        >

          <li
            class="cvi-Pagination-item cvi-Pagination-item--firstPage cvi-Pagination-item--disabled"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               ‹‹
            </a>
          </li>

          <li
            class="cvi-Pagination-item cvi-Pagination-item--previousPage cvi-Pagination-item--disabled"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               ‹
            </a>
          </li>


          <li
            class="cvi-Pagination-item cvi-Pagination-item--nextPage"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               ›
            </a>
          </li>

          <li
            class="cvi-Pagination-item cvi-Pagination-item--lastPage"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               ››
            </a>
          </li>
        </ul>
      </div>
    </cvi-pagination>
  </cvi-vision>
</ng-component>
`;

exports[`Pagination renders with pages in state 1`] = `
<ng-component
  testedWidget={[Function NgCviPagination]}
>
  <cvi-vision>
    <cvi-pagination>
      <div
        class="cvi-Pagination"
        ng-reflect-ng-class="cvi-Pagination,"
      >
        <ul
          class="cvi-Pagination-list"
        >

          <li
            class="cvi-Pagination-item cvi-Pagination-item--firstPage cvi-Pagination-item--disabled"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               ‹‹
            </a>
          </li>

          <li
            class="cvi-Pagination-item cvi-Pagination-item--previousPage cvi-Pagination-item--disabled"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               ‹
            </a>
          </li>

          <li
            class="cvi-Pagination-item cvi-Pagination-item--page cvi-Pagination-item--selected"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               1
            </a>
          </li>
          <li
            class="cvi-Pagination-item cvi-Pagination-item--page"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               2
            </a>
          </li>
          <li
            class="cvi-Pagination-item cvi-Pagination-item--page"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               3
            </a>
          </li>
          <li
            class="cvi-Pagination-item cvi-Pagination-item--page"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               4
            </a>
          </li>
          <li
            class="cvi-Pagination-item cvi-Pagination-item--page"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               5
            </a>
          </li>
          <li
            class="cvi-Pagination-item cvi-Pagination-item--page"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               6
            </a>
          </li>
          <li
            class="cvi-Pagination-item cvi-Pagination-item--page"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               7
            </a>
          </li>

          <li
            class="cvi-Pagination-item cvi-Pagination-item--nextPage"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               ›
            </a>
          </li>

          <li
            class="cvi-Pagination-item cvi-Pagination-item--lastPage"
          >
            <a
              class="cvi-Pagination-link"
              href="null"
            >
               ››
            </a>
          </li>
        </ul>
      </div>
    </cvi-pagination>
  </cvi-vision>
</ng-component>
`;
